packages/rich-text/src/component/use-input-and-selection.js

```diff
@@ -96,15 +97,53 @@ export function useInputAndSelection( props ) {
 				return;
 			}
 
-			const currentValue = createRecord();
-			const { start, activeFormats: oldActiveFormats = [] } =
+			const newValue = createRecord();
+			const { start: oldStart, activeFormats: oldActiveFormats = [] } =
 				record.current;
 
+			const SP = ' ';
+			const NBSP = '\u00a0';
+			const isSpOrNbsp = ( char ) => char === SP || char === NBSP;
+
+			// When inserting multiple spaces, alternate between a normal space
+			// and a non-breaking space. This is to prevent browsers from
+			// collapsing multiple spaces into one.
+			if (
+				isCollapsed( newValue ) &&
+				oldStart !== newValue.start &&
+				( oldStart < newValue.start
+					? newValue.text.slice( oldStart, newValue.start ) === SP
+					: isSpOrNbsp(
+							record.current.text.slice(
+								newValue.start,
+								oldStart
+							)
+					  ) )
+			) {
+				const text = newValue.text;
+				let startOffset = newValue.start;
+				let endOffset = newValue.start;
+
+				while ( isSpOrNbsp( text[ startOffset - 1 ] ) ) {
+					startOffset--;
+				}
+				while ( isSpOrNbsp( text[ endOffset ] ) ) {
+					endOffset++;
+				}
+
+				newValue.text =
+					text.slice( 0, startOffset ) +
+					Array.from( { length: endOffset - startOffset } )
+						.map( ( _, index ) => ( index % 2 === 0 ? SP : NBSP ) )
+						.join( '' ) +
+					text.slice( endOffset );
```

Comment:
would it make sense here to use a regular expression to handle this?

```js
// Replace all successive spaces with an alternating pattern of space/nbsp.
newValue.text = newValue.replace( /[ \u00a0]+/g, match => ''.padStart( match.length, ' \u00a0' ) );
```

Reply:
I don't think we want to be replacing _all_ spaces. We shouldn't modify existing content too much imo, only the spaces around the selection.